pkg/stemming/porter2english.go

```
@@ -0,0 +1,680 @@
+package stemming
+
+import (
+	"slices"
+	"strings"
+
+	"go.rtnl.ai/nlp/pkg/enum"
+	"go.rtnl.ai/nlp/pkg/shared"
+)
+
+// ############################################################################
+// Porter2Stemmer English Steps
+// ############################################################################
+
+// Returns the stem for the selected English word using the English Porter2
+// algorithm. Whitespace will be trimmed and the stem will be returned in
+// all lowercase.
+func (p *Porter2Stemmer) StemEnglish(word string) (stem string) {
+	// Ensure the language configured is English, in case the user did not call
+	// [Porter2Stemmer.Stem] to get here.
+	if p.lang != enum.LanguageEnglish {
```

Do you think this should return an error rather than changing the language, to avoid unexpected behavior? It would be even better if we could make this method impossible to be called with languages other than English, but I'm not sure how difficult that change would be.

---

I kinda figured that if someone calls StemEnglish then they know they want to stem an English word.

The reason I do not return an error here or from Stem is I believe that there should be no errors returned from such a function, since it should just stem the word and return the stem.

Mostly the reason why I set the language here, then reset it with a defer, is to ensure no problems are encountered due to many switch/case expressions that are peppered throughout the Porter2Stemmer code.

@pdeziel For now, I will leave it as-is due to those reasons, but we can revisit this later if you think we should!

---

I agree with your logic there; it reminds me of John Ousterhout's principle of [Define Errors Out of Existence](https://wiki.tcl-lang.org/page/Define+Errors+Out+of+Existence) which advocates for returning as few errors as possible. The idea is that returning errors often introduces more complexity than needed, which you pointed out.

I think the behavior works for now as long as it's documented appropriately for the caller of this function.